nimble_scanner: rework to enable scanning BLE5 PHYs and extended advertisements

haukepetersen opened the pull request:
### Contribution description
This PR makes the `nimble_scanner` more powerful by adding the means to also scan for extended advertisements and also for scanning using the coded PHY mode. Furthermore, the new design wraps transparently on top of both NimBLEs extended- and non-extended advertising interfaces, whichever is compiled in. 

NimBLE offers two different sets of APIs, depending if extended advertisements are enabled or not (value of `MYNEWT_VAL_BLE_EXT_ADV `). Once extended advertisements are enabled, the non-extended functions are still callable, but simply return a not-supported error. So once this option is enabled, the extended APIs have to be used for legacy as well as extended advertisements. As using the extended advertising features in NimBLE (`MYNEWT_VAL_BLE_EXT_ADV := 1`) adds significant ROM usage to the build, it is desirable for the scanner to work with both configurations, so users can select between extended features and reduced code size. By having the `nimble_scanner` design decoupled from the underlying NimBLE APIs, we allow users to switch between both modes without having to touch their calls to the `nimble_scanner` API.

To achieve this however, this PR breaks the existing API in some points. I think this is well worth it, as the API changes make the API more robust and future proof. Also this is not one of our high-impact APIs...

### Testing procedure
- compile and run the NimBLE scanner examples (`examples/nimble_scanner`) as-is -> should scan legacy advertisings just fine
- compile the scanner example with `USEMODULE="nimble_phy_coded nimble_phy_2mbit" (best on a nrf52840-based board) and verify that it also picks up extended advertisements

With BLE5 nodes around and extended advertisements enabled, the scan result will look something like this:
```
2021-09-13 21:44:46,098 # scan 2000                                                                                                                                                                                                            
2021-09-13 21:44:46,100 # Scanning for 2000 ms now ...                                                                                                                                                                                         
2021-09-13 21:44:48,102 # Done, results:                                                                                                                                                                                                       
2021-09-13 21:44:48,113 # [ 0] 00:04:2f:8e:3a:0f (PUBLIC) [EXT-CONN-SCAN] phy:1M-N/A "2;0:04:2F:8E:3A:0F;eTRV", adv_msg_cnt: 1, adv_int: 0us, last_rssi: -84                                                                                   
2021-09-13 21:44:48,123 # [ 1] 88:c6:26:8f:8e:36 (PUBLIC) [EXT-CONN-SCAN] phy:1M-N/A "", adv_msg_cnt: 7, adv_int: 258387us, last_rssi: -89                                                                                                     
2021-09-13 21:44:48,134 # [ 2] 57:4c:42:f5:b6:e0 (PUBLIC) [EXT-CONN-SCAN] phy:1M-N/A "TL100_B6E0", adv_msg_cnt: 59, adv_int: 33571us, last_rssi: -52                                                                                           
2021-09-13 21:44:48,153 # [ 3] d8:0f:99:30:b8:be (PUBLIC) [EXT-SCAN] phy:1M-N/A "undefined", adv_msg_cnt: 28, adv_int: 70198us, last_rssi: -79                                                                                                 
2021-09-13 21:44:48,165 # [ 4] c8:d0:83:ec:1b:e6 (PUBLIC) [EXT-CONN-SCAN] phy:1M-N/A "undefined", adv_msg_cnt: 10, adv_int: 109683us, last_rssi: -54                                                                                           
2021-09-13 21:44:48,178 # [ 5] 7c:ad:30:72:98:69 (RANDOM) [EXT-CONN-SCAN] phy:1M-N/A "undefined", adv_msg_cnt: 41, adv_int: 45193us, last_rssi: -65                                                                                            
2021-09-13 21:44:48,184 # [ 6] 29:91:a7:51:9c:4e (RANDOM) [NONCONN_IND] phy:1M-1M "undefined", adv_msg_cnt: 6, adv_int: 307346us, last_rssi: -65                                                                                               
2021-09-13 21:44:48,190 # [ 7] 46:06:42:60:1e:8a (RANDOM) [EXT-SCAN] phy:1M-N/A "undefined", adv_msg_cnt: 3, adv_int: 561542us, last_rssi: -76                                                                                                 
2021-09-13 21:44:48,198 # [ 8] 7c:8a:66:3f:ed:c5 (RANDOM) [EXT-CONN-SCAN] phy:1M-N/A "42750CMI", adv_msg_cnt: 5, adv_int: 249503us, last_rssi: -87                                                                                             
2021-09-13 21:44:48,209 # [ 9] 2a:1e:80:4f:94:2e (RANDOM) [NONCONN_IND] phy:1M-1M "undefined", adv_msg_cnt: 8, adv_int: 207876us, last_rssi: -86                                                                                               
2021-09-13 21:44:48,222 # [10] 65:e5:2a:30:37:06 (RANDOM) [EXT-CONN-SCAN] phy:1M-N/A "undefined", adv_msg_cnt: 17, adv_int: 95695us, last_rssi: -77                                                                                            
2021-09-13 21:44:48,230 # [11] 7b:3c:39:6c:94:cf (RANDOM) [EXT-CONN-SCAN] phy:1M-N/A "undefined", adv_msg_cnt: 3, adv_int: 181036us, last_rssi: -87                                                                                            
2021-09-13 21:44:48,241 # [12] a4:30:7a:85:77:fb (PUBLIC) [EXT-CONN-SCAN] phy:1M-N/A "undefined", adv_msg_cnt: 1, adv_int: 0us, last_rssi: -86                                                                                                 
2021-09-13 21:44:48,252 # [13] 60:d7:ed:41:09:78 (RANDOM) [EXT-CONN-SCAN] phy:1M-N/A "undefined", adv_msg_cnt: 1, adv_int: 0us, last_rssi: -78
2021-09-13 21:44:48,262 # [14] 1c:1a:c0:9b:55:65 (PUBLIC) [EXT-CONN-SCAN] phy:1M-N/A "undefined", adv_msg_cnt: 6, adv_int: 152604us, last_rssi: -82
2021-09-13 21:44:48,273 # [15] db:e4:f4:c2:21:da (RANDOM) [EXT-CONN] phy:CODED-CODED "horstus", adv_msg_cnt: 5, adv_int: 240250us, last_rssi: -54
2021-09-13 21:44:48,284 # [16] 40:78:7a:04:4b:02 (RANDOM) [EXT-CONN-SCAN] phy:1M-N/A "undefined", adv_msg_cnt: 6, adv_int: 138475us, last_rssi: -80
2021-09-13 21:44:48,294 # [17] 08:17:98:bd:3e:e0 (RANDOM) [NONCONN_IND] phy:1M-1M "undefined", adv_msg_cnt: 6, adv_int: 151663us, last_rssi: -56
2021-09-13 21:44:48,306 # [18] 00:04:2f:8e:3a:a6 (PUBLIC) [EXT-CONN-SCAN] phy:1M-N/A "2;0:04:2F:8E:3A:A6;eTRV", adv_msg_cnt: 1, adv_int: 0us, last_rssi: -87
2021-09-13 21:44:48,316 # [19] 04:9f:cf:c4:5d:8b (RANDOM) [NONCONN_IND] phy:1M-1M "undefined", adv_msg_cnt: 1, adv_int: 0us, last_rssi: -81
```


### Issues/PRs references
~~rebased on top of/waiting for https://github.com/RIOT-OS/RIOT/pull/16839~~ #16839 is merged

haukepetersen commented:
Forgot to mention: there are some more code parts that need adaption. But I wanted to get the initial approval first before I add those...

haukepetersen commented on pkg/nimble/Makefile.include:
```diff
@@ -77,12 +77,35 @@ endif
+
+ifneq (,$(filter nimble_adv_ext,$(USEMODULE)))
+  CFLAGS += -DMYNEWT_VAL_BLE_EXT_ADV=1
+  CFLAGS += -DMYNEWT_VAL_BLE_LL_EXT_ADV_AUX_PTR_CNT=2
```
Just checked, the default value for `MYNEWT_VAL_BLE_LL_SCAN_AUX_SEGMENT_CNT` is set to 8 in the RIOT specific  configuration (`porting/npl/riot/include/syscfg/syscfg.h`) -> so no need to adjust this here...

haukepetersen commented:
Added a test application that always uses extended advertisements and also pulls in the 2M and CODED PHY modes. So the build test should now cover both aspects of the scanner module: the new test the extended advertisings, and the exsiting `nimble_scanner` example the legacy code paths.

haukepetersen commented:
Pulled in two more fixes that were so far only pushed to #16860. I accidentally already squashed them right in, sorry :-)

haukepetersen commented:
Added two more fixes:
- by updating the NimBLE version, we pull in a change where the `ble_l2cap_ping` function was removed from NimBLE again, so we need to remove `nimble_netif_ping()` as well
- adapted the `sc_nimble_netif` shell command to the scanner API change

fjmolinas reviewed (commented):
Works as advertised, 

```
2021-12-02 17:39:49,852 # main(): This is RIOT! (Version: 2022.01-devel-871-g50348-pr-16843)
2021-12-02 17:39:49,853 # NimBLE Scanner Example Application
2021-12-02 17:39:49,853 # Type `scan help` for more information
> scan
2021-12-02 17:39:54,129 # scan
2021-12-02 17:39:54,132 # Scanning for 1000 ms now ...
2021-12-02 17:39:55,134 # Done, results:
2021-12-02 17:39:55,143 # [ 0] 52:98:0E:CC:43:E0 (random)  [IND] phy:1M-1M "undefined", adv_msg_cnt: 4, adv_int: 70577us, last_rssi: -80
2021-12-02 17:39:55,153 # [ 1] 69:BC:56:52:05:06 (random)  [SCAN_IND] phy:1M-1M "undefined", adv_msg_cnt: 8, adv_int: 96056us, last_rssi: -77
2021-12-02 17:39:55,164 # [ 2] F9:08:38:4F:27:50 (random)  [NONCONN_IND] phy:1M-1M "undefined", adv_msg_cnt: 1, adv_int: 0us, last_rssi: -79
2021-12-02 17:39:55,164 #
```

Can you add the following patch so it also works with `mynewt-core`

```diff
diff --git a/pkg/nimble/npl/include/nimble/nimble_npl_os.h b/pkg/nimble/npl/include/nimble/nimble_npl_os.h
index b223388f2c..811c856f3a 100644
--- a/pkg/nimble/npl/include/nimble/nimble_npl_os.h
+++ b/pkg/nimble/npl/include/nimble/nimble_npl_os.h
@@ -25,6 +25,10 @@
 #include "os/os.h"
 #include "mcu/mcu.h"

+#if defined(CPU_FAM_NRF51) || defined(CPU_FAM_NRF52)
+#include "nrf_clock.h"
+#endif
+
 #ifdef __cplusplus
 extern "C" {
 #endif
@@ -564,6 +568,20 @@ static inline void ble_npl_hw_set_isr(int irqn, void (*addr)(void))
     nrf5x_hw_set_isr(irqn, addr);
 }

+#if defined(CPU_FAM_NRF51) || defined(CPU_FAM_NRF52)
+static inline void
+nrf52_clock_hfxo_request(void)
+{
+    clock_hfxo_request();
+}
+
+static inline void
+nrf52_clock_hfxo_release(void)
+{
+    clock_hfxo_release();
+}
+#endif
+
 #ifdef __cplusplus
 }
 #endif

```

fjmolinas commented:
> Works as advertised

After adding the patch can you enable ci?

fjmolinas commented:
I guess my only comment is that this feature could use some documentation, but we can address it in #16860 since that PR introduces the use of multiple advertisement instances witch I'm quite interested in having well laid out.

haukepetersen commented:
Did add the patch you asked for, hope `mynewt-core` works now as it should with the current NimBLE master!

fjmolinas commented:
> Did add the patch you asked for, hope `mynewt-core` works now as it should with the current NimBLE master!

Works like a charm:

<details><summary>USEPKG=mynewt-core BOARD=dwm1001 make -C tests/nimble_ext_adv/ flash term -j</summary>

```
Building application "tests_nimble_ext_adv" for "dwm1001" with MCU "nrf52".

[INFO] cloning mynewt-core
Cloning into '/home/francisco/workspace/RIOT/build/pkg/mynewt-core'...
remote: Enumerating objects: 124451, done.
remote: Counting objects: 100% (3133/3133), done.
remote: Compressing objects: 100% (1370/1370), done.
remote: Total 124451 (delta 1650), reused 2747 (delta 1521), pack-reused 121318
Receiving objects: 100% (124451/124451), 172.34 MiB | 14.76 MiB/s, done.
Resolving deltas: 100% (76253/76253), done.
HEAD is now at de203365f hal: Update syscfg for BSPs with own copy of _sbrk
[INFO] updating mynewt-core /home/francisco/workspace/RIOT/build/pkg/mynewt-core/.pkg-state.git-downloaded
echo de203365f207dda658657a7525253e02f68503a1 > /home/francisco/workspace/RIOT/build/pkg/mynewt-core/.pkg-state.git-downloaded
[INFO] patch mynewt-core
[INFO] cloning nimble
Cloning into '/home/francisco/workspace/RIOT/build/pkg/nimble'...
remote: Enumerating objects: 43483, done.
remote: Counting objects: 100% (2403/2403), done.
remote: Compressing objects: 100% (1001/1001), done.
remote: Total 43483 (delta 1534), reused 1920 (delta 1319), pack-reused 41080
Receiving objects: 100% (43483/43483), 13.21 MiB | 16.40 MiB/s, done.
Resolving deltas: 100% (26635/26635), done.
HEAD is now at bc142016 host/mesh: fix loopback in net
[INFO] updating nimble /home/francisco/workspace/RIOT/build/pkg/nimble/.pkg-state.git-downloaded
echo bc142016bdef082ba7997aea09aec1c1e4db0104 > /home/francisco/workspace/RIOT/build/pkg/nimble/.pkg-state.git-downloaded
[INFO] patch nimble
[INFO] cloning tinycrypt
Cloning into '/home/francisco/workspace/RIOT/build/pkg/tinycrypt'...
remote: Enumerating objects: 698, done.
remote: Total 698 (delta 0), reused 0 (delta 0), pack-reused 698
Receiving objects: 100% (698/698), 891.88 KiB | 7.31 MiB/s, done.
Resolving deltas: 100% (458/458), done.
HEAD is now at 5969b0e Merge pull request #40 from winnietwo/side_channel_patch
[INFO] updating tinycrypt /home/francisco/workspace/RIOT/build/pkg/tinycrypt/.pkg-state.git-downloaded
echo 5969b0e0f572a15ed95dc272e57104faeb5eb6b0 > /home/francisco/workspace/RIOT/build/pkg/tinycrypt/.pkg-state.git-downloaded
[INFO] patch tinycrypt
"make" -C /home/francisco/workspace/RIOT/pkg/mynewt-core
"make" -C /home/francisco/workspace/RIOT/pkg/nimble
"make" -C /home/francisco/workspace/RIOT/pkg/tinycrypt
"make" -C /home/francisco/workspace/RIOT/pkg/nimble/addr/
"make" -C /home/francisco/workspace/RIOT/build/pkg/nimble/nimble/controller/src/ -f /home/francisco/workspace/RIOT/Makefile.base MODULE=nimble_controller
"make" -C /home/francisco/workspace/RIOT/build/pkg/nimble/nimble/drivers/nrf52/src/ -f /home/francisco/workspace/RIOT/Makefile.base MODULE=nimble_drivers_nrf5x
"make" -C /home/francisco/workspace/RIOT/build/pkg/nimble/nimble/host/src/ -f /home/francisco/workspace/RIOT/pkg/nimble/nimble.host.mk
"make" -C /home/francisco/workspace/RIOT/build/pkg/nimble/nimble/host/store/ram/src/ -f /home/francisco/workspace/RIOT/Makefile.base MODULE=nimble_host_store_ram
"make" -C /home/francisco/workspace/RIOT/build/pkg/nimble/nimble/host/util/src/ -f /home/francisco/workspace/RIOT/Makefile.base MODULE=nimble_host_util
"make" -C /home/francisco/workspace/RIOT/build/pkg/mynewt-core/hw/mcu/nordic/nrf52xxx/src/ -f /home/francisco/workspace/RIOT/pkg/mynewt-core/mynewt-core_nrf5x_hal.mk MODULE=mynewt-core_nrf5x_hal
"make" -C /home/francisco/workspace/RIOT/build/pkg/mynewt-core/kernel/os/src -f /home/francisco/workspace/RIOT/pkg/mynewt-core/mynewt-core_os.mk MODULE=mynewt-core_os
"make" -C /home/francisco/workspace/RIOT/build/pkg/mynewt-core/util/mem/src -f /home/francisco/workspace/RIOT/pkg/mynewt-core/mynewt-core_util.mk MODULE=mynewt-core_util
"make" -C /home/francisco/workspace/RIOT/build/pkg/nimble/porting/nimble/src/ -f /home/francisco/workspace/RIOT/pkg/nimble/nimble.porting.mk MODULE=nimble_porting_nimble
"make" -C /home/francisco/workspace/RIOT/build/pkg/tinycrypt/lib/source/ -f /home/francisco/workspace/RIOT/pkg/tinycrypt/Makefile.tinycrypt -f /home/francisco/workspace/RIOT/Makefile.base MODULE=tinycrypt
"make" -C /home/francisco/workspace/RIOT/pkg/nimble/contrib/
"make" -C /home/francisco/workspace/RIOT/pkg/nimble/scanlist
"make" -C /home/francisco/workspace/RIOT/pkg/nimble/scanner
"make" -C /home/francisco/workspace/RIOT/build/pkg/nimble/nimble/transport/ram/src/ -f /home/francisco/workspace/RIOT/Makefile.base MODULE=nimble_transport_ram
"make" -C /home/francisco/workspace/RIOT/boards/dwm1001
"make" -C /home/francisco/workspace/RIOT/core
"make" -C /home/francisco/workspace/RIOT/cpu/nrf52
"make" -C /home/francisco/workspace/RIOT/drivers
"make" -C /home/francisco/workspace/RIOT/pkg/mynewt-core/contrib
"make" -C /home/francisco/workspace/RIOT/sys
"make" -C /home/francisco/workspace/RIOT/cpu/cortexm_common
"make" -C /home/francisco/workspace/RIOT/cpu/nrf52/periph
"make" -C /home/francisco/workspace/RIOT/cpu/nrf52/vectors
"make" -C /home/francisco/workspace/RIOT/cpu/nrf5x_common
"make" -C /home/francisco/workspace/RIOT/drivers/periph_common
"make" -C /home/francisco/workspace/RIOT/sys/auto_init
"make" -C /home/francisco/workspace/RIOT/sys/event
"make" -C /home/francisco/workspace/RIOT/sys/fmt
"make" -C /home/francisco/workspace/RIOT/sys/frac
"make" -C /home/francisco/workspace/RIOT/sys/isrpipe
"make" -C /home/francisco/workspace/RIOT/sys/malloc_thread_safe
"make" -C /home/francisco/workspace/RIOT/cpu/cortexm_common/periph
"make" -C /home/francisco/workspace/RIOT/sys/net/ble/bluetil
"make" -C /home/francisco/workspace/RIOT/sys/newlib_syscalls_default
"make" -C /home/francisco/workspace/RIOT/cpu/nrf5x_common/periph
"make" -C /home/francisco/workspace/RIOT/sys/sema
"make" -C /home/francisco/workspace/RIOT/sys/shell
"make" -C /home/francisco/workspace/RIOT/sys/net/ble/bluetil/bluetil_ad
"make" -C /home/francisco/workspace/RIOT/sys/shell/commands
"make" -C /home/francisco/workspace/RIOT/sys/net/ble/bluetil/bluetil_addr
"make" -C /home/francisco/workspace/RIOT/sys/stdio_uart
"make" -C /home/francisco/workspace/RIOT/sys/tsrb
"make" -C /home/francisco/workspace/RIOT/sys/ztimer
   text	   data	    bss	    dec	    hex	filename
  92076	    272	  12964	 105312	  19b60	/home/francisco/workspace/RIOT/tests/nimble_ext_adv/bin/dwm1001/tests_nimble_ext_adv.elf
/home/francisco/workspace/RIOT/dist/tools/jlink/jlink.sh flash /home/francisco/workspace/RIOT/tests/nimble_ext_adv/bin/dwm1001/tests_nimble_ext_adv.bin
### Flashing Target ###
### Flashing at base address 0x0 with offset 0 ###
SEGGER J-Link Commander V7.52a (Compiled Jul 28 2021 11:12:35)
DLL version V7.52a, compiled Jul 28 2021 11:12:19

J-Link Commander will now exit on Error

J-Link Command File read successfully.
Processing script file...

J-Link connection not established yet but required for command.
Connecting to J-Link via USB...O.K.
Firmware: J-Link OB-STM32F072-128KB-CortexM compiled Feb  2 2021 16:45:08
Hardware version: V1.00
S/N: 760103435
VTref=3.300V
Target connection not established yet but required for command.
Device "NRF52" selected.


Connecting to target via SWD
InitTarget() start
InitTarget() end
Found SW-DP with ID 0x2BA01477
DPIDR: 0x2BA01477
Scanning AP map to find all available APs
AP[2]: Stopped AP scan as end of AP map has been reached
AP[0]: AHB-AP (IDR: 0x24770011)
AP[1]: JTAG-AP (IDR: 0x02880000)
Iterating through AP map to find AHB-AP to use
AP[0]: Core found
AP[0]: AHB-AP ROM base: 0xE00FF000
CPUID register: 0x410FC241. Implementer code: 0x41 (ARM)
Found Cortex-M4 r0p1, Little endian.
FPUnit: 6 code (BP) slots and 2 literal slots
CoreSight components:
ROMTbl[0] @ E00FF000
ROMTbl[0][0]: E000E000, CID: B105E00D, PID: 000BB00C SCS-M7
ROMTbl[0][1]: E0001000, CID: B105E00D, PID: 003BB002 DWT
ROMTbl[0][2]: E0002000, CID: B105E00D, PID: 002BB003 FPB
ROMTbl[0][3]: E0000000, CID: B105E00D, PID: 003BB001 ITM
ROMTbl[0][4]: E0040000, CID: B105900D, PID: 000BB9A1 TPIU
ROMTbl[0][5]: E0041000, CID: B105900D, PID: 000BB925 ETM
Cortex-M4 identified.
Halting CPU for downloading file.
Downloading file [/home/francisco/workspace/RIOT/tests/nimble_ext_adv/bin/dwm1001/tests_nimble_ext_adv.bin]...
Comparing flash   [100%] Done.
Erasing flash     [100%] Done.
Programming flash [100%] Done.
J-Link: Flash download: Bank 0 @ 0x00000000: 1 range affected (94208 bytes)
J-Link: Flash download: Total: 1.863s (Prepare: 0.089s, Compare: 0.007s, Erase: 0.050s, Program & Verify: 1.658s, Restore: 0.056s)
J-Link: Flash download: Program & Verify speed: 55 KiB/s
O.K.

Reset delay: 0 ms
Reset type NORMAL: Resets core & peripherals via SYSRESETREQ & VECTRESET bit.
Reset: Halt core after reset via DEMCR.VC_CORERESET.
Reset: Reset device via AIRCR.SYSRESETREQ.



Script processing completed.

/home/francisco/workspace/RIOT/dist/tools/pyterm/pyterm -p "/dev/riot/tty-dwm1001" -b "115200"  
Twisted not available, please install it if you want to use pyterm's JSON capabilities
2021-12-02 22:06:33,808 # Connect to serial port /dev/riot/tty-dwm1001
Welcome to pyterm!
Type '/exit' to exit.
2021-12-02 22:06:34,812 # main(): This is RIOT! (Version: 2022.01-devel-872-gb6e214-pr-16843)
2021-12-02 22:06:34,813 # NimBLE Scanner Example Application
2021-12-02 22:06:34,813 # Type `scan help` for more information
> scan
2021-12-02 22:06:37,173 # scan
2021-12-02 22:06:37,175 # Scanning for 1000 ms now ...
2021-12-02 22:06:38,177 # Done, results:
2021-12-02 22:06:38,186 # [ 0] 4F:AD:0D:EF:5C:6F (random)  [IND] phy:1M-1M "undefined", adv_msg_cnt: 6, adv_int: 91773us, last_rssi: -84
2021-12-02 22:06:38,196 # [ 1] 6A:70:E7:41:C6:D1 (random)  [IND] phy:1M-1M "undefined", adv_msg_cnt: 6, adv_int: 91250us, last_rssi: -82
2021-12-02 22:06:38,206 # [ 2] 7D:30:CA:A0:BC:B4 (random)  [IND] phy:1M-1M "undefined", adv_msg_cnt: 1, adv_int: 0us, last_rssi: -85
2021-12-02 22:06:38,216 # [ 3] 69:88:24:2E:4F:6A (random)  [NONCONN_IND] phy:1M-1M "undefined", adv_msg_cnt: 2, adv_int: 310316us, last_rssi: -85
2021-12-02 22:06:38,227 # [ 4] 71:40:D8:A8:10:D2 (random)  [SCAN_IND] phy:1M-1M "undefined", adv_msg_cnt: 5, adv_int: 102199us, last_rssi: -77
2021-12-02 22:06:38,227 # 
```

</details>

fjmolinas commented on cpu/nrf52/Makefile.features:
```diff
@@ -25,4 +25,11 @@ FEATURES_PROVIDED += cortexm_mpu
 
 FEATURES_PROVIDED += periph_i2c_reconfigure
 
+# On top of the default 1Mbit PHY mode, all nrf52 support the 2MBit PHY mode,
+# and the 52840 does further support the coded PHYs
+FEATURES_PROVIDED += ble_phy_2mbit
+ifneq (,$(filter nrf52840% nrf52811%,$(CPU_MODEL)))
+  FEATURES_PROVIDED += ble_phy_coded
+endif
```
I forgot can you add this to Kconfig otherwise we will have the ci fail here

haukepetersen commented:
Fixed/adapted two more things:
- added the coded PHY feature also to the nRF52820 and nRF52833 CPU models
- exposed and mapped the PHY feature to Kconfig

haukepetersen commented:
And as always one more thing that was missed: adapted `nimble_autoconn` and `nimble_rpble` to scanner API changes.

fjmolinas reviewed (approved):
ACK!

haukepetersen commented:
Awesome, thanks for the review!
